lib/src/Pos/Logic/Types.hs

```diff
@@ -101,11 +100,11 @@ data Logic m = Logic
     , securityParams     :: SecurityParams
     }
 
--- | We have to hoist a pipes producer, so the Monad constraint arises.
+-- | We have to hoist a Conduit producer, so the Monad constraint arises.
 hoistLogic :: Monad m => (forall x . m x -> n x) -> Logic m -> Logic n
 hoistLogic nat logic = logic
     { getSerializedBlock = nat . getSerializedBlock logic
-    , streamBlocks = unsafeHoist nat . streamBlocks logic
+    , streamBlocks = transPipe nat . streamBlocks logic
```

Comment:
❗️`transPipe` comes with [a warning](https://www.stackage.org/haddock/lts-11.7/conduit-1.3.0.2/Data-Conduit.html#v:transPipe) in the docs:

> Note that the monad transforming function will be run multiple times, resulting in unintuitive behavior in some cases. For a fuller treatment, please see:
> 
> https://github.com/snoyberg/conduit/wiki/Dealing-with-monad-transformers

From reading the github post *I think* we're ok, because [the only use of `hoistLogic`](https://github.com/input-output-hk/cardano-sl/blob/develop/lib/src/Pos/Launcher/Runner.hs#L98) [calls `elimRealMode` to unwrap a ReaderT](https://github.com/input-output-hk/cardano-sl/blob/develop/lib/src/Pos/Launcher/Runner.hs#L115), which shouldn't involve the state-carry-over effects mentioned as potential problems.

Reply:
Yep, that seems legit!

Reply:
Another way of saying this is that the passed in function `foo :: (forall x. m x -> n x)` must be a _monad morphism_ and not just any natural transformation. What this means, more or less, is that `foo a >> foo b` should behave the same as `foo (a >> b)`. If we set `foo = flip evalState 1`, it's easy to see how these two examples will behave differently.  I would recommend improving the documentation here, saying that the given function is expected to be a /monad morphism/.

Reply:
@k0001 Thanks, that is a good recommendation. I've updated the comment [here](https://github.com/input-output-hk/cardano-sl/blob/mhuesch/CDEC-464/lib/src/Pos/Logic/Types.hs#L104).